### ISSUE TITLE
Scripts requested via systemjs are rejected with 401 unauthorized

### ISSUE THREAD

**Issue by akiander** [systemjs/systemjs#1150]
I [also posted this to stackoverflow](http://stackoverflow.com/questions/35899799/scripts-requested-via-systemjs-fail-with-401-unauthorized), but trying to cross-reference it here as well. 

I have a web application that runs on Team Foundation Server 2015. The application works fine as-is, but I want to convert it over to use systemjs. 

When I attempt to do that, I can see that systemjs is requesting the root javascript module correctly, and the module is there, but the call to load it fails with a 401 unauthorized. 

The issue is not with authentication- the same code works fine and can successfully load the module without using systemjs, so I know that something related to using systemjs to load the script is the problem. 

Any ideas on how to solve something like this?

**Comment by guybedford** [comment]
I've responded to the SO post.

**Comment by akiander** [comment]
This fixed my issue - so thank you! 

But as soon as the script was loaded successfully, I then got this issue: 

Uncaught TypeError: Unexpected anonymous System.register call.
(anonymous function) @ system.src.js:2885
(anonymous function) @ system.src.js:3532
(anonymous function) @ system.src.js:4005
(anonymous function) @ system.src.js:2608
SystemJSLoader.register @ system.src.js:2841
(anonymous function) @ main.js:1

It happens on the very first line of my script, which is the call to System.Register() 

I can see it [happens to others](https://github.com/systemjs/systemjs/issues/861)

I tried using globalEvaluationScope: false in my config but that did not work. 

Any ideas?

**Comment by guybedford** [comment]
What web browser are you using?

**Comment by akiander** [comment]
Chrome

**Comment by guybedford** [comment]
Is there anyway you can share a replication? The only thing I can think is if the security policy is stopping the ability from registering the script load completion properly.

**Comment by guybedford** [comment]
Also ensure the `System.register(` is directly in the outer scope of execution and not wrapped or anything like that.

**Comment by akiander** [comment]
This application is hosted as a page (aka hub) inside of Microsoft's Team Foundation Server 2015. 

I created a dummy account that you can use to reproduce the issue. 

The page that fails [is here](https://testingsystemjs.visualstudio.com/DefaultCollection/My%20First%20Project/_apps/hub/akiander.source-extensions.akiander.SourceHub?hubGroupId=ms.vss-web.home-hub-group)

The credentials are: 

     ACCOUNT:        testing.systemjs@outlook.com
     PASSWORD:      testing@system.js

You should see that the page displays "Loading..." and in the console log, you should see the error (Uncaught TypeError: Unexpected anonymous System.register call) that I reference above.

Let me know if you need anything else.

**Comment by guybedford** [comment]
Sure I had a look and SystemJS is definitely not fetching `main.js` in that example, because if you add a `console.log('fetching ' + load.name)` to line 2640 of SystemJS it doesn't log anything to show that System.import is actually doing the fetch. So I'm not sure what is doing the request to that file here for you.

**Comment by akiander** [comment]
My systemjs configuration has `scriptLoad: true` and so the `sourceHub.html` has the following script tag that loads scripts/main.js on line 45:  
	
	`<script src="scripts/main.js"></script>`

That's how main.js gets loaded. Does that answer your question?

**Comment by guybedford** [comment]
SystemJS needs to do the loading for you, so you should remove that script tag.

**Comment by akiander** [comment]
That's what [I had originally](http://stackoverflow.com/questions/35899799/scripts-requested-via-systemjs-fail-with-401-unauthorized) but when systemjs does the loading, then the main.js file is rejected with a 401 unauthorized. 

You mentioned that systemjs uses XHR to load scripts... is there a way to influence systemjs to load the script file [with credentials](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/withCredentials)? 

Maybe that would solve my original issue?

**Comment by guybedford** [comment]
@akiander `scriptLoad: true` will switch SystemJS from XHR to doing its own script tag injection. Otherwise you can do XHR with credentials as well with `authorization: true`.

**Comment by akiander** [comment]
I noticed that `scriptLoad: true` wasn't there so I added it back in. And I also added `authorization: true` and that resulting implementation appears to work now. Here's how I have it configured

      System.config({
        authorization: true,
        packages: {
          scripts: {
            format: 'register',
            defaultExtension: 'js',
            scriptLoad: true
          }
        },
        meta: {
			'scripts/*': { crossOrigin: 'use-credentials' }
		}
      });
      System.import('scripts/main')
            .then(null, console.error.bind(console));

Did I put the `authorization: true` in the correct location? I didn't see it in specified on the [configuration options](https://github.com/systemjs/systemjs/blob/master/docs/config-api.md) page.

**Comment by nervgh** [comment]
It works for me with SystemJS v0.20.5
```js
SystemJS.config({
  meta: {
    '*': {
      authorization: true
    }
});
```

**Comment by thanikc** [comment]
I see in the [docs/config-api](https://github.com/systemjs/systemjs/blob/master/docs/config-api.md): 

> [authorization]: This can be a custom authorization header string for XHR requests made by SystemJS.

This makes more sense than `authorization: true`. The only problem is: I can't get it to work.

I've tried 
`meta: { 
  '*': { authorization: 'Basic dGVzdDp0ZXN0' }
}`

No Authorization header was sent. I've tried putting the value into an array (or what are the brackets around [authorization] for?) -> still no header. How is this meant to be used? An example would be nice.
